src/bin/pg_autoctl/supervisor.c

```diff
@@ -56,22 +72,277 @@ static bool supervisor_restart_service(Supervisor *supervisor,
+	memcpy(dynamicService, service, sizeof(*service)); /* IGNORE-BANNED */
+	supervisor->dynamicServices.serviceCount++;
+
+	(void) unblock_signals(&sig_mask_orig);
+
+	return true;
+}
+
+
+/*
+ * supervisor_remove_dynamic stops a running dynamic service with serviceName.
+ * Once the service is stopped, then it is removed from the dynamicService array
+ * and supervisor_loop can largely forget about it.
+ *
+ * However, it will get notified during waitpid, so we add this service to the
+ * recentlyRemoved array and we will hold it there until it is requested via
+ * supervisor_find_recently_removed_dynamic() which subsequently removes it.
+ *
+ * Returns true if the service is stopped and removed from the dynamicService
+ * array, false otherwise.
+ */
+bool
+supervisor_remove_dynamic(Supervisor *supervisor, const char *serviceName)
+{
+	Service *dynamicService = NULL;
+	int serviceCount = supervisor->dynamicServices.serviceCount;
+	int serviceIndex;
+
+	/* Find it in the dynamic array */
+	for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+	{
+		Service *service = &supervisor->dynamicServices.services[serviceIndex];
+
+		if (!strncmp(service->name, serviceName, strlen(serviceName)))
+		{
+			dynamicService = service;
+			break;
+		}
+	}
+
+	/* It was not found, nothing to do. */
+	if (!dynamicService)
+	{
+		return false;
+	}
+
+	/*
+	 * Since we are going to signal the service, we have to make certain that we
+	 * will not error out in the supervisor_loop on the waitpid() response. So
+	 * after killing the service and removing it from the dynamicService array,
+	 * we add it to the recentlyRemoved array to find it again later.
+	 */
+	if (!kill(dynamicService->pid, SIGTERM))
+	{
+		log_info("Service %s with %d did not receive signal",
+				 dynamicService->name, dynamicService->pid);
+	}
+
+	log_info("Stopped pg_autoctl %s service with pid %d",
+			 dynamicService->name, dynamicService->pid);
+
+	/*
+	 * Remove the dynamic Service from the array by swapping it out with the
+	 * last element in the array and bringing the array count down by one
+	 */
+	supervisor->dynamicServices.services[serviceIndex] =
+		supervisor->dynamicServices.services[serviceCount - 1];
+	supervisor->dynamicServices.serviceCount--;
+
+	/*
+	 * Finally add the dynamic service to the recently removed array.
+	 *
+	 * If we reached the end of this array, start again at the beginning. This
+	 * is a loosely held array and we do not fuss too much.
+	 */
+	if (supervisor->dynamicRecentlyRemoved.serviceCount == MAXDYNSERVICES)
+	{
+		supervisor->dynamicRecentlyRemoved.serviceCount = 0;
+	}
+
+	/*
+	 * IGNORE-BANNED fits here because we want to take ownership of the service.
+	 * We copy to an address that we own from a non overlapping address.
+	 */
+	memcpy(&supervisor->dynamicRecentlyRemoved.services[    /* IGNORE-BANNED */
+			   supervisor->dynamicRecentlyRemoved.serviceCount],
+		   dynamicService,
+		   sizeof(*dynamicService));
```

Comment:
I there a better way to just copy the immediate structure instance there? We just use the C assignment in many places in the code for that, when we know that our structures are constructed without pointers to other places in memory.

Reply:
Same as above, please point me to the code you suggest and I will happily change

Reply:
Have a look at https://github.com/citusdata/pg_auto_failover/blob/master/src/bin/pg_autoctl/keeper_config.c#L304 and many other such places in the code for examples of how we like to copy data from a structure instance to another in this project ;-)